runtime/src/impls.rs

```diff
@@ -269,6 +260,198 @@ where
 	}
 }
 
+/// CENNZnet implementation of the evm runner which handles the case where users are attempting
+/// to set their payment asset. In this case, we will exchange their desired asset into CPAY to
+/// complete the transaction
+pub struct FeePreferencesRunner<T>
+where
+	T: pallet_evm::Config + crml_eth_wallet::Config,
+{
+	_marker: PhantomData<T>,
+}
+
+impl<T> FeePreferencesRunner<T>
+where
+	T: pallet_evm::Config + crml_eth_wallet::Config<AccountId = AccountId>,
+{
+	/// Decodes the input for call_with_fee_preferences
+	fn decode_input(
```

Comment:
for testing sake, I'd make a new error type for the cases here. The caller can make the decision to map_err it to WithdrawFailed

Reply:
same for the calculate_total_gas